pkg/daemon/drain.go

```diff
@@ -45,34 +45,30 @@ func (dn *Daemon) cordonOrUncordonNode(desired bool) error {
 }
 
 func (dn *Daemon) drain() error {
-	backoff := wait.Backoff{
-		Steps:    5,
-		Duration: 10 * time.Second,
-		Factor:   2,
+	drainer := func() chan error {
+		ret := make(chan error)
+		go func() {
+			for {
+				if err := drain.RunNodeDrain(dn.drainer, dn.node.Name); err != nil {
+					glog.Infof("Draining failed with: %v, retrying", err)
+					time.Sleep(5 * time.Minute)
+					continue
+				}
+				close(ret)
+				return
+			}
+		}()
+		return ret
 	}
-	var lastErr error
-	if err := wait.ExponentialBackoff(backoff, func() (bool, error) {
-		err := drain.RunNodeDrain(dn.drainer, dn.node.Name)
-		if err != nil {
-			lastErr = err
-			glog.Infof("Draining failed with: %v, retrying", err)
-			return false, nil
-		}
-		return true, nil
-
-	}); err != nil {
-		if err == wait.ErrWaitTimeout {
-			failMsg := fmt.Sprintf("%d tries: %v", backoff.Steps, lastErr)
-			MCDDrainErr.WithLabelValues(dn.node.Name, "WaitTimeout").Set(float64(backoff.Steps))
-			dn.recorder.Eventf(getNodeRef(dn.node), corev1.EventTypeWarning, "FailedToDrain", failMsg)
-			return errors.Wrapf(lastErr, "failed to drain node (%d tries): %v", backoff.Steps, err)
-		}
-		MCDDrainErr.WithLabelValues(dn.node.Name, "UnknownError").Set(float64(backoff.Steps))
-		dn.recorder.Eventf(getNodeRef(dn.node), corev1.EventTypeWarning, "FailedToDrain", err.Error())
-		return errors.Wrap(err, "failed to drain node")
+	select {
+	case <-time.After(24 * time.Hour):
```

Comment:
I'm still relatively new to go, but it looks like if this error case is triggered nothing is going to stop `drainer()`. Is that right, or is there something subtle that I'm not seeing?

Reply:
that was uhh.. not right :)

Reply:
This looks better. :-) I would probably move line 76 up to happen first in this block, but the order may not actually matter.

Reply:
sure, gave it a whirl :smile: